lib/workflows/nfs-volumes/index.js

```diff
@@ -244,38 +251,200 @@ function provisionNfsVolumes(job, cb) {
+                         * it already exists. If we get an error loading the
+                         * volume then, we'll make the task fail.
+                         */
+                        next();
                     }
+                });
+            },
+            function reserve(ctx, next) {
+                if (job.params.vm_uuid === undefined) {
+                    job.params.vm_uuid = uuid.v4();
+                }
 
-                    job.log.info({loadedVolume: loadedVolume},
-                        'Loaded existing volume');
-                    callback(listVolumesErr, loadedVolume);
-                    return;
+                job.log.info({
+                    volume_uuid: ctx.volumeUuid,
+                    job_uuid: job.uuid,
+                    vm_uuid: job.params.vm_uuid,
+                    owner_uuid: job.params.owner_uuid
+                }, 'Reserving volume');
+
+                volapi.reserveVolume({
+                    owner_uuid: job.params.owner_uuid,
+                    job_uuid: job.uuid,
+                    vm_uuid: job.params.vm_uuid,
+                    volume_name: volumeName
+                }, function onVolRes(volResErr, volRes) {
+                    ctx.volumeReservation = volRes;
+                    next(volResErr);
                 });
-            } else {
-                job.log.info({createdVolume: createdVolume},
-                    'Created new volume');
-                callback(volumeCreationErr, createdVolume);
-            }
-        });
+            },
+            function provision(ctx, next) {
+                var volumeCreationParams;
+
+                if (ctx.volume) {
+                    job.log.info({
+                        volume: ctx.volume
+                    }, 'Volume already exits, no need to create it');
+                    next();
+                    return;
+                }
+
+                job.log.info('Volume does not exists, creating it');
+
+                volumeCreationParams = {
+                    name: volumeName,
+                    owner_uuid: job.params.owner_uuid,
+                    type: 'tritonnfs'
+                };
+
+                /*
+                 * If the volume doesn't exist, then we created its uuid
+                 * beforehand to register a reservation for it, so we must
+                 * pass that uuid to VOLAPI so that it uses that uuid when
+                 * creating the volume to match the reservation.
+                 */
+                if (ctx.volumeUuid !== undefined) {
+                    volumeCreationParams.uuid = ctx.volumeUuid;
+                }
+
+                volumeCreationParams.networks =
+                    vmNics.map(function getOverlayNetwork(network) {
+                        job.log.info({network: network}, 'network');
+
+                        if (typeof (network) !== 'object') {
+                            callback('network must be an object');
```

Comment:
> If somehow we had more than one non-object in the vmNics, wouldn't this mean we'll call callback() multiple times?

Yes, good catch!

> Would it make sense to assert instead on

I don't want to assert in workflow jobs because that basically prevents any other task in the workflow from being executed, including the onerror tasks chain.

So instead I'll refactor this code to call the callback once if there's 1 or more nics that are not of the expected type/layout. How does that sound?